File: ios/TimberCruisingApp/Screens/TiltBadge.swift
// Live device-tilt badge for the Diameter scan screen.
//
// The single biggest source of DBH bias is a leaning phone — a chord
// projected through a non-vertical slice of the cylinder reads a
// systematically wrong diameter. Showing the cruiser the live pitch
// + roll lets them self-correct before tapping. Arboreal Forest
// shipped this in v4.98 ("improved visualisation: device inclination").
//
// Self-contained on purpose: owns its own CMMotionManager so any
// scan screen can drop one in without plumbing IMU through its
// view model. Pairs with GPSAccuracyBadge in the same `topStrip`.
//
// Tier mapping uses the existing confidence palette so the colour
// language matches the rest of the app:
//   • |pitch| ≤ 3°      → Good   (confidenceOk)
//   • |pitch| ≤ 8°      → Fair   (confidenceWarn)
//   • |pitch|  > 8°     → Check  (confidenceBad)

import SwiftUI

#if canImport(CoreMotion) && os(iOS)
import CoreMotion
#endif

public struct TiltBadge: View {

    @StateObject private var motion = TiltMonitor()

    public init() {}

    public var body: some View {
        let tier = currentTier
        return HStack(spacing: 6) {
            Image(systemName: "level")
                .font(.system(size: 11, weight: .semibold))
                .foregroundStyle(tier.color)
            Text(tier.label)
                .font(ForestixType.dataSmall)
                .foregroundStyle(.white)
            if let p = motion.pitchDeg {
                Text(String(format: "%+.0f°", p))
                    .font(ForestixType.dataSmall)
                    .foregroundStyle(.white.opacity(0.75))
            }
        }
        .padding(.horizontal, 10)
        .padding(.vertical, 6)
        .background(Color.black.opacity(0.55))
        .overlay(
            Capsule().stroke(Color.white.opacity(0.20), lineWidth: 0.5))
        .clipShape(Capsule())
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("Device tilt \(tier.label.lowercased())")
        .accessibilityValue(motion.pitchDeg.map {
            String(format: "%.0f degrees", $0)
        } ?? "no reading")
        .onAppear { motion.start() }
        .onDisappear { motion.stop() }
    }

    private struct Tier {
        let label: String
        let color: Color
    }

    private var currentTier: Tier {
        guard let p = motion.pitchDeg else {
            return Tier(label: "Tilt —", color: ForestixPalette.confidenceBad)
        }
        let abs = Swift.abs(p)
        if abs <= 3 {
            return Tier(label: "Level", color: ForestixPalette.confidenceOk)
        } else if abs <= 8 {
            return Tier(label: "Tilted", color: ForestixPalette.confidenceWarn)
        } else {
            return Tier(label: "Tilted", color: ForestixPalette.confidenceBad)
        }
    }
}

// MARK: - Live pitch source

#if canImport(CoreMotion) && os(iOS)

@MainActor
private final class TiltMonitor: ObservableObject {

    @Published private(set) var pitchDeg: Double?

    private let manager = CMMotionManager()
    private let queue = OperationQueue()

    init() {
        queue.name = "com.forestix.tilt"
        queue.qualityOfService = .userInitiated
    }

    func start() {
        guard manager.isDeviceMotionAvailable else { return }
        manager.deviceMotionUpdateInterval = 1.0 / 30.0  // 30 Hz; UI doesn't need 100
        manager.startDeviceMotionUpdates(
            using: .xArbitraryZVertical,
            to: queue
        ) { [weak self] data, _ in
            guard let self, let d = data else { return }
            // Pitch from gravity vector — same convention as
            // IMUHelpers.pitchFromGravity. Returns radians; convert
            // to degrees for display.
            let g = d.gravity
            // pitch = atan2(-g.z, sqrt(g.x² + g.y²)) is the
            // device's nose-up/down angle. Use the same sign
            // convention the height pipeline uses.
            let pitch = atan2(-g.z, (g.x * g.x + g.y * g.y).squareRoot())
            let degrees = pitch * 180 / .pi
            Task { @MainActor [weak self] in
                self?.pitchDeg = degrees
            }
        }
    }

    func stop() {
        manager.stopDeviceMotionUpdates()
    }
}

#else

@MainActor
private final class TiltMonitor: ObservableObject {
    @Published private(set) var pitchDeg: Double?
    func start() {}
    func stop() {}
}

#endif

File: android/app/src/main/java/com/hcjeong/forestix/ui/screens/TiltBadge.kt
// Live device-tilt badge for the Diameter scan screen — port of iOS
// Screens/TiltBadge.swift.
//
// The single biggest source of DBH bias is a leaning phone — a chord
// projected through a non-vertical slice of the cylinder reads a
// systematically wrong diameter. Showing the cruiser the live pitch lets
// them self-correct before tapping.
//
// Pitch source: the AR camera's forward elevation (ArController), the same
// convention the height pipeline uses — so the badge, the dev HUD, and the
// research CSV all agree on sign.
//
// Tier mapping uses the existing confidence palette so the colour language
// matches the rest of the app:
//   • |pitch| ≤ 3°  → Level  (confidenceOk)
//   • |pitch| ≤ 8°  → Tilted (confidenceWarn)
//   • |pitch| > 8°  → Tilted (confidenceBad)

package com.hcjeong.forestix.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Straighten
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.hcjeong.forestix.ar.ArController
import com.hcjeong.forestix.ui.theme.Forestix
import kotlinx.coroutines.delay
import java.util.Locale

@Composable
fun TiltBadge(controller: ArController, modifier: Modifier = Modifier) {
    val colors = Forestix.colors
    val type = Forestix.type
    var pitchDeg by remember { mutableStateOf<Float?>(null) }
    LaunchedEffect(Unit) {
        while (true) {
            pitchDeg = controller.cameraPitchDeg()
            delay(100)
        }
    }
    val p = pitchDeg
    val (label, tint) = when {
        p == null -> "Tilt —" to colors.confidenceBad
        kotlin.math.abs(p) <= 3f -> "Level" to colors.confidenceOk
        kotlin.math.abs(p) <= 8f -> "Tilted" to colors.confidenceWarn
        else -> "Tilted" to colors.confidenceBad
    }
    Row(
        modifier
            .clip(CircleShape)
            .background(Color.Black.copy(alpha = 0.55f))
            .border(0.5.dp, Color.White.copy(alpha = 0.20f), CircleShape)
            .padding(horizontal = 10.dp, vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        // Bubble-level glyph — nearest Material analogue of SF "level".
        Icon(
            Icons.Filled.Straighten,
            contentDescription = null,
            tint = tint,
            modifier = Modifier.size(11.dp),
        )
        Text(label, style = type.dataSmall, color = Color.White)
        if (p != null) {
            Text(
                String.format(Locale.US, "%+.0f°", p),
                style = type.dataSmall.copy(fontSize = 13.sp),
                color = Color.White.copy(alpha = 0.75f),
            )
        }
    }
}
